From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sun, 12 Feb 2023 10:10:22 +0930
Subject: [PATCH] Fix pardoning uncached game profiles


diff --git a/src/main/java/org/bukkit/craftbukkit/ban/CraftProfileBanList.java b/src/main/java/org/bukkit/craftbukkit/ban/CraftProfileBanList.java
index 89ad99df2f76537bdcb0482f2046e1215c104ded..19efeda7debaddf424be659f84d84dd875bfda19 100644
--- a/src/main/java/org/bukkit/craftbukkit/ban/CraftProfileBanList.java
+++ b/src/main/java/org/bukkit/craftbukkit/ban/CraftProfileBanList.java
@@ -155,6 +155,19 @@ public class CraftProfileBanList implements ProfileBanList {
     public void pardon(String target) {
         Preconditions.checkArgument(target != null, "Target cannot be null");
 
+        // Nabulus start
+        if (CraftProfileBanList.getProfile(target) == null) {
+            java.util.List<NameAndId> toPardon = new java.util.ArrayList<>();
+            for (UserBanListEntry banListEntry : list.getEntries()) {
+                if (banListEntry == null || banListEntry.getUser() == null) continue;
+                if (banListEntry.getUser().id().toString().equalsIgnoreCase(target)) {
+                    toPardon.add(banListEntry.getUser());
+                }
+            }
+            toPardon.forEach(this.list::remove);
+            return;
+        }
+        // Nabulus end
         this.pardon(CraftProfileBanList.getProfile(target));
     }
 
