From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sun, 29 Jun 2025 09:35:02 +0930
Subject: [PATCH] Dispatch Discord commands through CommandMap


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index b2961752b50e98d95adf3af4f658e2fdb2127d43..614a585cf7383ac1db381d3a9010e753fdb508cb 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -924,6 +924,11 @@ public final class CraftServer implements Server {
         Preconditions.checkArgument(rawSender != null, "sender cannot be null");
         Preconditions.checkArgument(commandLine != null, "commandLine cannot be null");
         org.spigotmc.AsyncCatcher.catchOp("Command Dispatched Async: " + commandLine); // Spigot // Paper - Include command in error message
+        // Nabulus start
+        if (rawSender instanceof org.bukkit.OfflinePlayer && rawSender instanceof org.bukkit.command.ConsoleCommandSender) {
+            return this.commandMap.dispatch(rawSender, commandLine);
+        }
+        // Nabulus end
         CommandSourceStack sourceStack = VanillaCommandWrapper.getListener(rawSender);
 
         String command = StringUtils.normalizeSpace(commandLine.trim());
