From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sun, 29 Jun 2025 09:35:02 +0930
Subject: [PATCH] Dispatch Discord commands through CommandMap


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 3187d470e62da15e8a1f60fc1bd4ee213bf2f69d..b59727fb69c0454fd12a99dc0d7c369eee180544 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -925,6 +925,11 @@ public final class CraftServer implements Server {
         Preconditions.checkArgument(rawSender != null, "sender cannot be null");
         Preconditions.checkArgument(commandLine != null, "commandLine cannot be null");
         org.spigotmc.AsyncCatcher.catchOp("Command Dispatched Async: " + commandLine); // Spigot // Paper - Include command in error message
+        // Nabulus start
+        if (rawSender instanceof org.bukkit.OfflinePlayer && rawSender instanceof org.bukkit.command.ConsoleCommandSender) {
+            return this.commandMap.dispatch(rawSender, commandLine);
+        }
+        // Nabulus end
         CommandSourceStack sourceStack = VanillaCommandWrapper.getListener(rawSender);
 
         String command = StringUtils.normalizeSpace(commandLine.trim());
