From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sun, 29 Jun 2025 09:35:02 +0930
Subject: [PATCH] Dispatch Discord commands through CommandMap


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index bfdf63b42ccfb6728fb99fa7ba45ee413b86d77a..961a933046fba252fe2e4d691ae16deb4609399d 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -977,6 +977,11 @@ public final class CraftServer implements Server {
         Preconditions.checkArgument(rawSender != null, "sender cannot be null");
         Preconditions.checkArgument(commandLine != null, "commandLine cannot be null");
         org.spigotmc.AsyncCatcher.catchOp("Command Dispatched Async: " + commandLine); // Spigot // Paper - Include command in error message
+        // Nabulus start
+        if (rawSender instanceof org.bukkit.OfflinePlayer && rawSender instanceof org.bukkit.command.ConsoleCommandSender) {
+            return this.commandMap.dispatch(rawSender, commandLine);
+        }
+        // Nabulus end
         CommandSourceStack sourceStack = VanillaCommandWrapper.getListener(rawSender);
 
         net.minecraft.commands.Commands commands = this.getHandle().getServer().getCommands();
