From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sun, 27 Feb 2022 21:27:01 +0930
Subject: [PATCH] Separate creative levels


diff --git a/ca/spottedleaf/dataconverter/minecraft/versions/V1022.java b/ca/spottedleaf/dataconverter/minecraft/versions/V1022.java
index dab97581780d0d3156335fe0fbb84ea0823c5585..4917c74145220fda494bcc17382ba2a9436507d1 100644
--- a/ca/spottedleaf/dataconverter/minecraft/versions/V1022.java
+++ b/ca/spottedleaf/dataconverter/minecraft/versions/V1022.java
@@ -17,6 +17,7 @@ public final class V1022 {
 
             WalkerUtils.convertList(MCTypeRegistry.ITEM_STACK, data, "Inventory", fromVersion, toVersion);
             WalkerUtils.convertList(MCTypeRegistry.ITEM_STACK, data, "EnderItems", fromVersion, toVersion);
+            WalkerUtils.convertList(MCTypeRegistry.ITEM_STACK, data, "Nabulus.CreativeLevelInventory", fromVersion, toVersion);
 
             WalkerUtils.convert(MCTypeRegistry.ENTITY, data, "ShoulderEntityLeft", fromVersion, toVersion);
             WalkerUtils.convert(MCTypeRegistry.ENTITY, data, "ShoulderEntityRight", fromVersion, toVersion);
diff --git a/ca/spottedleaf/dataconverter/minecraft/versions/V4312.java b/ca/spottedleaf/dataconverter/minecraft/versions/V4312.java
index 7015d97cb5b97d59ca41f63ba23da1edfb6bae0a..98999e673ba29329948cbba8838fc1f6ed419175 100644
--- a/ca/spottedleaf/dataconverter/minecraft/versions/V4312.java
+++ b/ca/spottedleaf/dataconverter/minecraft/versions/V4312.java
@@ -27,7 +27,14 @@ public final class V4312 {
 
             @Override
             public MapType convert(final MapType data, final long sourceVersion, final long toVersion) {
-                final ListType inventory = data.getListUnchecked("Inventory");
+                // Nabulus start
+                convertInventory(data, "Inventory", "equipment");
+                convertInventory(data, "Nabulus.CreativeLevelInventory", "Nabulus.CreativeLevelEquipment");
+                return null;
+            }
+            private Void convertInventory(final MapType data, String inventoryDataKey, String equipmentDataKey) {
+                final ListType inventory = data.getListUnchecked(inventoryDataKey);
+                // Nabulus end
                 if (inventory == null) {
                     return null;
                 }
@@ -53,7 +60,7 @@ public final class V4312 {
                     equipment.setMap(equipmentKey, item);
                 }
 
-                data.setMap("equipment", equipment);
+                data.setMap(equipmentDataKey, equipment); // Nabulus
 
                 return null;
             }
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 785045f119ee2f9b9df4efc6e2f10134a1cbb160..a4de03a71324c22774546ff1607284b51356d95f 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -647,6 +647,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         }
         this.setLevel(level);
         this.gameMode.setLevel(level);
+        this.chooseInventory(); // Nabulus
     }
     // CraftBukkit end
 
@@ -1545,6 +1546,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                 profilerFiller.pop();
                 profilerFiller.push("placing");
                 this.setServerLevel(level);
+                this.chooseInventory(); // Nabulus
                 this.connection.internalTeleport(PositionMoveRotation.of(teleportTransition), teleportTransition.relatives()); // CraftBukkit - use internal teleport without event
                 this.connection.resetPosition();
                 level.addDuringTeleport(this);
@@ -2147,11 +2149,21 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
 
     public void restoreFrom(ServerPlayer that, boolean keepEverything) {
+        // Nabulus start
+        this.restoreFrom(that, keepEverything, false);
+    }
+    public void restoreFrom(ServerPlayer that, boolean keepEverything, boolean toOrFromCreativeLevel) {
+        // Nabulus end
         this.wardenSpawnTracker = that.wardenSpawnTracker;
         this.chatSession = that.chatSession;
         this.gameMode.setGameModeForPlayer(that.gameMode.getGameModeForPlayer(), that.gameMode.getPreviousGameModeForPlayer());
         this.onUpdateAbilities();
         this.getAttributes().assignBaseValues(that.getAttributes());
+        // Nabulus start
+        this.normalInventory.replaceWith(that.normalInventory);
+        this.creativeLevelInventory.replaceWith(that.creativeLevelInventory);
+        this.usingNormalInventory = that.usingNormalInventory;
+        // Nabulus end
         if (keepEverything) {
             // this.getAttributes().assignPermanentModifiers(that.getAttributes()); // CraftBukkit
             this.setHealth(that.getHealth());
@@ -2180,6 +2192,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.seenCredits = that.seenCredits;
         this.enteredNetherPosition = that.enteredNetherPosition;
         this.chunkTrackingView = that.chunkTrackingView;
+        if (toOrFromCreativeLevel) return; // Nabulus
         this.requestedDebugSubscriptions = that.requestedDebugSubscriptions;
         this.setShoulderEntityLeft(that.getShoulderEntityLeft());
         this.setShoulderEntityRight(that.getShoulderEntityRight());
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 989ac565c47a70c7947cb7315d0f5c2cfecd0363..9a3cc199963d1e593218c6c294c6fedb5b45d0c4 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -606,8 +606,15 @@ public abstract class PlayerList {
         player.level().removePlayerImmediately(player, reason);
         ServerLevel level = teleportTransition.newLevel();
         ServerPlayer serverPlayer = player; // Paper - TODO - recreate instance
+        // Nabulus start
+        boolean toOrFromCreativeLevel = ((net.minecraft.world.level.storage.PrimaryLevelData) teleportTransition.newLevel().levelData).creativeLevel || ((net.minecraft.world.level.storage.PrimaryLevelData) player.level().levelData).creativeLevel;
+        if (toOrFromCreativeLevel) {
+            serverPlayer.releaseLeftShoulderEntity();
+            serverPlayer.releaseRightShoulderEntity();
+        }
+        // Nabulus end
         serverPlayer.connection = player.connection;
-        serverPlayer.restoreFrom(player, keepInventory);
+        serverPlayer.restoreFrom(player, keepInventory, toOrFromCreativeLevel); // Nabulus
         serverPlayer.setId(player.getId());
         serverPlayer.setMainArm(player.getMainArm());
         if (false && !teleportTransition.missingRespawnBlock()) { // Paper - Once we not reuse the player entity, this can be flipped again but without the events being fired
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 21d0de44d9055a5e80c551402abdb7e352dcb0c9..ba63f48654e5c599ad752067e8aa71233e7301db 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4000,6 +4000,11 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         }
         // Paper end - Fix item duplication and teleport issues
         if (this.level() instanceof ServerLevel serverLevel && !this.isRemoved()) {
+            // Nabulus start
+            if (!this.level.equals(serverLevel) && ((net.minecraft.world.level.storage.PrimaryLevelData) this.level().levelData).creativeLevel) {
+                return null;
+            }
+            // Nabulus end
             // CraftBukkit start
             PositionMoveRotation absolutePosition = PositionMoveRotation.calculateAbsolute(PositionMoveRotation.of(this), PositionMoveRotation.of(teleportTransition), teleportTransition.relatives());
             Vec3 velocity = absolutePosition.deltaMovement(); // Paper
diff --git a/net/minecraft/world/entity/PortalProcessor.java b/net/minecraft/world/entity/PortalProcessor.java
index 4e69e32ad000b8c02e44c3ffce24ff5e3c090a2f..cb0ffb5e2be2bd2d4e850f9dd8af7d91890167b3 100644
--- a/net/minecraft/world/entity/PortalProcessor.java
+++ b/net/minecraft/world/entity/PortalProcessor.java
@@ -29,6 +29,11 @@ public class PortalProcessor {
     }
 
     public @Nullable TeleportTransition getPortalDestination(ServerLevel level, Entity entity) {
+        // Nabulus start
+        if (((net.minecraft.world.level.storage.PrimaryLevelData) entity.level().levelData).creativeLevel) {
+            return null;
+        }
+        // Nabulus end
         return this.portal.getPortalDestination(level, entity, this.entryPosition);
     }
 
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 7bf160cef8bd772d27d4a71f04d5d9861ef67984..33f5954591b0ccff532e420f02471b5c0f4f74da 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -184,6 +184,36 @@ public abstract class Player extends Avatar implements ContainerUser {
     public boolean fauxSleeping;
     public int oldLevel = -1;
 
+    // Nabulus start
+    protected final Inventory normalInventory = new Inventory(this, this.createEquipment());
+    protected final Inventory creativeLevelInventory = new Inventory(this, this.createEquipment());
+    public boolean usingNormalInventory = true;
+
+    public void chooseInventory() {
+        if (((net.minecraft.world.level.storage.PrimaryLevelData)this.level().levelData).creativeLevel) {
+            if (usingNormalInventory) {
+                this.normalInventory.replaceWith(this.inventory);
+                this.inventory.replaceWith(this.creativeLevelInventory);
+                usingNormalInventory = false;
+            }
+        } else {
+            if (!usingNormalInventory) {
+                this.creativeLevelInventory.replaceWith(this.inventory);
+                this.inventory.replaceWith(this.normalInventory);
+                usingNormalInventory = true;
+            }
+        }
+    }
+
+    public Inventory getNormalInventory() {
+        return usingNormalInventory ? this.inventory : this.normalInventory;
+    }
+
+    public Inventory getCreativeLevelInventory() {
+        return usingNormalInventory ? this.creativeLevelInventory : this.inventory;
+    }
+    // Nabulus end
+
     @Override
     public org.bukkit.craftbukkit.entity.CraftHumanEntity getBukkitEntity() {
         return (org.bukkit.craftbukkit.entity.CraftHumanEntity) super.getBukkitEntity();
@@ -669,6 +699,14 @@ public abstract class Player extends Avatar implements ContainerUser {
         this.setUUID(this.gameProfile.id());
         this.inventory.load(input.listOrEmpty("Inventory", ItemStackWithSlot.CODEC));
         this.inventory.setSelectedSlot(input.getIntOr("SelectedItemSlot", 0));
+        // Nabulus start
+        this.normalInventory.replaceWith(this.inventory);
+        this.creativeLevelInventory.load(input.listOrEmpty("Nabulus.CreativeLevelInventory", ItemStackWithSlot.CODEC));
+        this.creativeLevelInventory.setSelectedSlot(input.getIntOr("Nabulus.CreativeLevelSelectedItemSlot", 0));
+        this.creativeLevelInventory.equipment.setAll(input.read("Nabulus.CreativeLevelEquipment", EntityEquipment.CODEC).orElseGet(EntityEquipment::new));
+        this.usingNormalInventory = true;
+        chooseInventory();
+        // Nabulus end
         this.sleepCounter = input.getShortOr("SleepTimer", (short)0);
         this.experienceProgress = input.getFloatOr("XpP", 0.0F);
         this.experienceLevel = input.getIntOr("XpLevel", 0);
@@ -691,10 +729,24 @@ public abstract class Player extends Avatar implements ContainerUser {
 
     @Override
     protected void addAdditionalSaveData(ValueOutput output) {
+        // Nabulus start
+        if (!usingNormalInventory) {
+            this.creativeLevelInventory.replaceWith(this.inventory);
+            this.inventory.replaceWith(this.normalInventory);
+        }
+        // Nabulus end
         super.addAdditionalSaveData(output);
         NbtUtils.addCurrentDataVersion(output);
         this.inventory.save(output.list("Inventory", ItemStackWithSlot.CODEC));
         output.putInt("SelectedItemSlot", this.inventory.getSelectedSlot());
+        // Nabulus start
+        this.creativeLevelInventory.save(output.list("Nabulus.CreativeLevelInventory", ItemStackWithSlot.CODEC));
+        output.putInt("Nabulus.CreativeLevelSelectedItemSlot", this.creativeLevelInventory.getSelectedSlot());
+        output.store("Nabulus.CreativeLevelEquipment", EntityEquipment.CODEC, this.creativeLevelInventory.equipment);
+        if (!usingNormalInventory) {
+            this.inventory.replaceWith(this.creativeLevelInventory);
+        }
+        // Nabulus end
         output.putShort("SleepTimer", (short)this.sleepCounter);
         output.putFloat("XpP", this.experienceProgress);
         output.putInt("XpLevel", this.experienceLevel);
diff --git a/net/minecraft/world/level/storage/LevelStorageSource.java b/net/minecraft/world/level/storage/LevelStorageSource.java
index 140630186c3b0324c248cc2a2f31d3b906557b9c..7d76ad1dca83d1be71f0c5bb32b726639b260c07 100644
--- a/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -143,6 +143,7 @@ public class LevelStorageSource {
             dynamic, levelSettings, complete.specialWorldProperty(), worldGenSettings.options(), lifecycle
         );
         primaryLevelData.pdc = (Tag) dynamic.getElement("BukkitValues", null); // CraftBukkit - Add PDC to world
+        primaryLevelData.creativeLevel = ((Dynamic<Tag>) dynamic).get("Nabulus.CreativeLevel").asBoolean(false); // Nabulus
         return new LevelDataAndDimensions(primaryLevelData, complete);
     }
 
diff --git a/net/minecraft/world/level/storage/PrimaryLevelData.java b/net/minecraft/world/level/storage/PrimaryLevelData.java
index 8936b4b1ae9c82eed2a7d3191a29f01792438d14..d136a99d0ffd8a0aab364dec9e7a9083b7fb5f99 100644
--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -77,6 +77,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
     public net.minecraft.core.Registry<net.minecraft.world.level.dimension.LevelStem> customDimensions;
     private net.minecraft.server.level.ServerLevel world;
     protected net.minecraft.nbt.Tag pdc;
+    public boolean creativeLevel;
 
     public void setWorld(net.minecraft.server.level.ServerLevel world) {
         if (this.world != null) {
@@ -293,6 +294,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
         tag.storeNullable("WanderingTraderId", UUIDUtil.CODEC, this.wanderingTraderId);
         tag.putString("Bukkit.Version", org.bukkit.Bukkit.getName() + "/" + org.bukkit.Bukkit.getVersion() + "/" + org.bukkit.Bukkit.getBukkitVersion()); // CraftBukkit
         this.world.getWorld().storeBukkitValues(tag); // CraftBukkit - add pdc
+        tag.putBoolean("Nabulus.CreativeLevel", this.creativeLevel); // Nabulus
     }
 
     private static ListTag stringCollectionToTag(Set<String> stringCollection) {
