From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Tue, 1 Mar 2022 19:44:48 +0930
Subject: [PATCH] Show hidden players in playerlist and tab suggestions


diff --git a/net/minecraft/commands/CommandSourceStack.java b/net/minecraft/commands/CommandSourceStack.java
index 6ead209287ab021b8245fa5c3b7f18ca7d9c66b3..8dd4c5afdab3f71c21b3b765a09a1b90f870c871 100644
--- a/net/minecraft/commands/CommandSourceStack.java
+++ b/net/minecraft/commands/CommandSourceStack.java
@@ -580,7 +580,7 @@ public class CommandSourceStack implements ExecutionCommandSource<CommandSourceS
 
     @Override
     public Collection<String> getOnlinePlayerNames() {
-        return this.entity instanceof ServerPlayer sourcePlayer && !sourcePlayer.getBukkitEntity().hasPermission("paper.bypass-visibility.tab-completion") ? this.getServer().getPlayerList().getPlayers().stream().filter(serverPlayer -> sourcePlayer.getBukkitEntity().canSee(serverPlayer.getBukkitEntity())).map(serverPlayer -> serverPlayer.getGameProfile().name()).toList() : Lists.newArrayList(this.server.getPlayerNames()); // Paper - Make CommandSourceStack respect hidden players
+        return this.entity instanceof ServerPlayer sourcePlayer && !sourcePlayer.getBukkitEntity().hasPermission("paper.bypass-visibility.tab-completion") ? this.getServer().getPlayerList().getPlayers().stream().filter(serverPlayer -> true || sourcePlayer.getBukkitEntity().canSee(serverPlayer.getBukkitEntity())).map(serverPlayer -> serverPlayer.getGameProfile().name()).toList() : Lists.newArrayList(this.server.getPlayerNames()); // Paper - Make CommandSourceStack respect hidden players // Nabulus - don't make it respect hidden players
     }
 
     @Override
diff --git a/net/minecraft/server/commands/ListPlayersCommand.java b/net/minecraft/server/commands/ListPlayersCommand.java
index e7c778a7f292fd0749cbd5e6586ad1b93ac7f29e..dac56a9c4e3ccacc0178878897e0a36b04138213 100644
--- a/net/minecraft/server/commands/ListPlayersCommand.java
+++ b/net/minecraft/server/commands/ListPlayersCommand.java
@@ -36,7 +36,7 @@ public class ListPlayersCommand {
         List<ServerPlayer> playersTemp = playerList.getPlayers();
         if (source.getBukkitSender() instanceof org.bukkit.entity.Player) {
             org.bukkit.entity.Player sender = (org.bukkit.entity.Player) source.getBukkitSender();
-            playersTemp = playersTemp.stream().filter((ep) -> sender.canSee(ep.getBukkitEntity())).collect(java.util.stream.Collectors.toList());
+            //playersTemp = playersTemp.stream().filter((ep) -> sender.canSee(ep.getBukkitEntity())).collect(java.util.stream.Collectors.toList()); // Nabulus
         }
         final List<ServerPlayer> players = playersTemp;
         // CraftBukkit end
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index a07568d90380f572a9eae68bc3fc1a60a9f69565..1e77e0904d3625db5a4092818d96aa7fe40a5875 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -253,7 +253,7 @@ public abstract class PlayerList {
         for (int i = 0; i < this.players.size(); ++i) {
             ServerPlayer entityplayer1 = (ServerPlayer) this.players.get(i);
 
-            if (entityplayer1.getBukkitEntity().canSee(bukkitPlayer)) {
+                if (entityplayer1.getBukkitEntity().canSee(bukkitPlayer) || true) { // Nabulus
                 // Paper start - Add Listing API for Player
                 if (entityplayer1.getBukkitEntity().isListed(bukkitPlayer)) {
                     // Paper end - Add Listing API for Player
@@ -266,7 +266,7 @@ public abstract class PlayerList {
             }
 
             if (entityplayer1 == player || !bukkitPlayer.canSee(entityplayer1.getBukkitEntity())) { // Paper - Use single player info update packet on join; Don't include joining player
-                continue;
+                    //continue; // Nabulus
             }
 
             onlinePlayers.add(entityplayer1); // Paper - Use single player info update packet on join
@@ -512,6 +512,7 @@ public abstract class PlayerList {
             if (otherPlayer.getBukkitEntity().canSee(player.getBukkitEntity())) {
                 otherPlayer.connection.send(packet);
             } else {
+                otherPlayer.connection.send(packet); // Nabulus
                 otherPlayer.getBukkitEntity().onEntityRemove(player);
             }
         }
