From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sat, 6 Aug 2022 13:21:50 +0930
Subject: [PATCH] Use system messages for player chat until Velocity support
 added


diff --git a/net/minecraft/network/chat/OutgoingChatMessage.java b/net/minecraft/network/chat/OutgoingChatMessage.java
index efcd5ba38738cb92d69bf95e476272e865fa7c43..a74bc79aac462d51189d45f9cb4259e4355c274e 100644
--- a/net/minecraft/network/chat/OutgoingChatMessage.java
+++ b/net/minecraft/network/chat/OutgoingChatMessage.java
@@ -28,6 +28,7 @@ public interface OutgoingChatMessage {
         @Override
         public void sendToPlayer(final ServerPlayer player, final boolean filtered, final ChatType.Bound chatType, final @org.jspecify.annotations.Nullable Component unsigned) {
             player.connection.sendDisguisedChatMessage(unsigned != null ? unsigned : this.content, chatType);
+            player.connection.sendDisguisedChatMessage(this.content, chatType); // Nabulus
             // Paper end
         }
     }
