From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sat, 6 Aug 2022 13:21:50 +0930
Subject: [PATCH] Use system messages for player chat until Velocity support
 added


diff --git a/net/minecraft/network/chat/OutgoingChatMessage.java b/net/minecraft/network/chat/OutgoingChatMessage.java
index 4635182c73f28f63ba84982cb6bcb422221b738e..19d87c84ebd788ff6ed09dfd1b45ddc8a9f8be64 100644
--- a/net/minecraft/network/chat/OutgoingChatMessage.java
+++ b/net/minecraft/network/chat/OutgoingChatMessage.java
@@ -27,6 +27,7 @@ public interface OutgoingChatMessage {
         }
         public void sendToPlayer(ServerPlayer player, boolean filtered, ChatType.Bound boundType, @javax.annotation.Nullable Component unsigned) {
             player.connection.sendDisguisedChatMessage(unsigned != null ? unsigned : this.content, boundType);
+            player.connection.sendDisguisedChatMessage(this.content, boundType); // Nabulus
             // Paper end
         }
     }
