From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sat, 6 Aug 2022 13:21:50 +0930
Subject: [PATCH] Use system messages for player chat until Velocity support
 added


diff --git a/src/main/java/net/minecraft/network/chat/OutgoingChatMessage.java b/src/main/java/net/minecraft/network/chat/OutgoingChatMessage.java
index c87b708c368713a23a10ad97704575ee4df27891..55c78d906c66b4d83bd6787ae323edab50e962a5 100644
--- a/src/main/java/net/minecraft/network/chat/OutgoingChatMessage.java
+++ b/src/main/java/net/minecraft/network/chat/OutgoingChatMessage.java
@@ -26,7 +26,7 @@ public interface OutgoingChatMessage {
             this.sendToPlayer(sender, filterMaskEnabled, params, null);
         }
         public void sendToPlayer(ServerPlayer sender, boolean filterMaskEnabled, ChatType.Bound params, @javax.annotation.Nullable Component unsigned) {
-            sender.connection.sendDisguisedChatMessage(unsigned != null ? unsigned : this.content, params);
+            sender.connection.sendDisguisedChatMessage(this.content, params); // Nabulus
             // Paper end
         }
     }
