From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Wed, 15 Nov 2023 17:15:10 +0930
Subject: [PATCH] Implement tick rate options for OneShot behaviors


diff --git a/net/minecraft/world/entity/ai/behavior/Behavior.java b/net/minecraft/world/entity/ai/behavior/Behavior.java
index b620d38e61ae560074b551a92afefe4022d5078e..e4dc3492b2e23a0f06ebc71905e968f56a0f6d26 100644
--- a/net/minecraft/world/entity/ai/behavior/Behavior.java
+++ b/net/minecraft/world/entity/ai/behavior/Behavior.java
@@ -16,6 +16,16 @@ public abstract class Behavior<E extends LivingEntity> implements BehaviorContro
     private final int minDuration;
     private final int maxDuration;
     private final String configKey; // Paper - configurable behavior tick rate and timings
+    // Nabulus start
+    public static String getConfigKey(String behaviorClass) {
+        String key = behaviorClass;
+        int lastSeparator = key.lastIndexOf('.');
+        if (lastSeparator != -1) {
+            key = key.substring(lastSeparator + 1);
+        }
+        return key.toLowerCase(java.util.Locale.ROOT);
+    }
+    // Nabulus end
 
     public Behavior(final Map<MemoryModuleType<?>, MemoryStatus> entryCondition) {
         this(entryCondition, DEFAULT_DURATION);
@@ -30,12 +40,7 @@ public abstract class Behavior<E extends LivingEntity> implements BehaviorContro
         this.maxDuration = maxDuration;
         this.entryCondition = entryCondition;
         // Paper start - configurable behavior tick rate and timings
-        String key = this.getClass().getName();
-        int lastSeparator = key.lastIndexOf('.');
-        if (lastSeparator != -1) {
-            key = key.substring(lastSeparator + 1);
-        }
-        this.configKey = key.toLowerCase(java.util.Locale.ROOT);
+        this.configKey = getConfigKey(this.getClass().getName()); // Nabulus
         // Paper end - configurable behavior tick rate and timings
     }
 
diff --git a/net/minecraft/world/entity/ai/behavior/OneShot.java b/net/minecraft/world/entity/ai/behavior/OneShot.java
index d7b97e35e1aa5264f191048cb448d18d28f75ea8..bacea4c085a54a0d6df80c3483783e80e69e6e92 100644
--- a/net/minecraft/world/entity/ai/behavior/OneShot.java
+++ b/net/minecraft/world/entity/ai/behavior/OneShot.java
@@ -7,6 +7,19 @@ import net.minecraft.world.entity.ai.behavior.declarative.Trigger;
 public abstract class OneShot<E extends LivingEntity> implements BehaviorControl<E>, Trigger<E> {
     private Behavior.Status status = Behavior.Status.STOPPED;
 
+    // Nabulus start - configurable behavior tick rate
+    private final String configKey;
+    private long lastRun;
+
+    public OneShot() {
+        this("unknown");
+    }
+
+    public OneShot(String configKey) {
+        this.configKey = configKey;
+    }
+    // Nabulus end
+
     @Override
     public final Behavior.Status getStatus() {
         return this.status;
@@ -14,8 +27,15 @@ public abstract class OneShot<E extends LivingEntity> implements BehaviorControl
 
     @Override
     public final boolean tryStart(final ServerLevel level, final E body, final long timestamp) {
+        // Nabulus start - behavior tick rate and timings
+        int tickRate = java.util.Objects.requireNonNullElse(level.paperConfig().tickRates.behavior.get(body.getType(), this.configKey), -1);
+        if (tickRate > -1 && timestamp < this.lastRun + tickRate) {
+            return false;
+        }
+        // Nabulus end
         if (this.trigger(level, body, timestamp)) {
             this.status = Behavior.Status.RUNNING;
+            this.lastRun = timestamp; // Nabulus - behavior tick rate
             return true;
         } else {
             return false;
diff --git a/net/minecraft/world/entity/ai/behavior/declarative/BehaviorBuilder.java b/net/minecraft/world/entity/ai/behavior/declarative/BehaviorBuilder.java
index 9cac66173a1a20e7bafca3a53cf84960c4806ccb..299f0b7f0816058f8bec48587dc7199c282c9561 100644
--- a/net/minecraft/world/entity/ai/behavior/declarative/BehaviorBuilder.java
+++ b/net/minecraft/world/entity/ai/behavior/declarative/BehaviorBuilder.java
@@ -40,7 +40,15 @@ public class BehaviorBuilder<E extends LivingEntity, M> implements App<BehaviorB
         final Function<BehaviorBuilder.Instance<E>, ? extends App<BehaviorBuilder.Mu<E>, Trigger<E>>> builder
     ) {
         final BehaviorBuilder.TriggerWithResult<E, Trigger<E>> resolvedBuilder = get((App<BehaviorBuilder.Mu<E>, Trigger<E>>)builder.apply(instance()));
-        return new OneShot<E>() {
+        // Nabulus start - configurable tick rates
+        java.lang.StackTraceElement[] stackTraceElements = Thread.currentThread().getStackTrace();
+        String callingClass = null;
+        if (stackTraceElements.length > 2) {
+            callingClass = stackTraceElements[2].getClassName();
+        }
+        final String configKey = callingClass == null ? "unknown" : net.minecraft.world.entity.ai.behavior.Behavior.getConfigKey(callingClass);
+        // Nabulus end
+        return new OneShot<E>(configKey) { // Nabulus
             @Override
             public boolean trigger(final ServerLevel level, final E body, final long timestamp) {
                 Trigger<E> trigger = resolvedBuilder.tryTrigger(level, body, timestamp);
