From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Wed, 15 Nov 2023 17:15:10 +0930
Subject: [PATCH] Implement tick rate options for OneShot behaviors


diff --git a/net/minecraft/world/entity/ai/behavior/Behavior.java b/net/minecraft/world/entity/ai/behavior/Behavior.java
index 5b0cadd2544fb2a627822e645ff32fec2e9cfda9..9a56c3a9dc97c5880bd6ac6f703d640fefb80be1 100644
--- a/net/minecraft/world/entity/ai/behavior/Behavior.java
+++ b/net/minecraft/world/entity/ai/behavior/Behavior.java
@@ -15,6 +15,16 @@ public abstract class Behavior<E extends LivingEntity> implements BehaviorContro
     private final int minDuration;
     private final int maxDuration;
     private final String configKey; // Paper - configurable behavior tick rate and timings
+    // Nabulus start
+    public static String getConfigKey(String behaviorClass) {
+        String key = io.papermc.paper.util.MappingEnvironment.reobf() ? io.papermc.paper.util.ObfHelper.INSTANCE.deobfClassName(behaviorClass) : behaviorClass;
+        int lastSeparator = key.lastIndexOf('.');
+        if (lastSeparator != -1) {
+            key = key.substring(lastSeparator + 1);
+        }
+        return key.toLowerCase(java.util.Locale.ROOT);
+    }
+    // Nabulus end
 
     public Behavior(Map<MemoryModuleType<?>, MemoryStatus> entryCondition) {
         this(entryCondition, 60);
@@ -29,12 +39,7 @@ public abstract class Behavior<E extends LivingEntity> implements BehaviorContro
         this.maxDuration = maxDuration;
         this.entryCondition = entryCondition;
         // Paper start - configurable behavior tick rate and timings
-        String key = io.papermc.paper.util.MappingEnvironment.reobf() ? io.papermc.paper.util.ObfHelper.INSTANCE.deobfClassName(this.getClass().getName()) : this.getClass().getName();
-        int lastSeparator = key.lastIndexOf('.');
-        if (lastSeparator != -1) {
-            key = key.substring(lastSeparator + 1);
-        }
-        this.configKey = key.toLowerCase(java.util.Locale.ROOT);
+        this.configKey = getConfigKey(this.getClass().getName()); // Nabulus
         // Paper end - configurable behavior tick rate and timings
     }
 
diff --git a/net/minecraft/world/entity/ai/behavior/OneShot.java b/net/minecraft/world/entity/ai/behavior/OneShot.java
index b43335515a89a5d6ab0edcf00e6447e25d8cd91c..ab6c5aaa63e0069c20e9db3529ddcc4bf8e0ad61 100644
--- a/net/minecraft/world/entity/ai/behavior/OneShot.java
+++ b/net/minecraft/world/entity/ai/behavior/OneShot.java
@@ -6,6 +6,18 @@ import net.minecraft.world.entity.ai.behavior.declarative.Trigger;
 
 public abstract class OneShot<E extends LivingEntity> implements BehaviorControl<E>, Trigger<E> {
     private Behavior.Status status = Behavior.Status.STOPPED;
+    // Nabulus start - configurable behavior tick rate
+    private final String configKey;
+    private long lastRun;
+
+    public OneShot() {
+        this("unknown");
+    }
+
+    public OneShot(String configKey) {
+        this.configKey = configKey;
+    }
+    // Nabulus end
 
     @Override
     public final Behavior.Status getStatus() {
@@ -14,8 +26,15 @@ public abstract class OneShot<E extends LivingEntity> implements BehaviorControl
 
     @Override
     public final boolean tryStart(ServerLevel level, E entity, long gameTime) {
+        // Nabulus start - behavior tick rate and timings
+        int tickRate = java.util.Objects.requireNonNullElse(level.paperConfig().tickRates.behavior.get(entity.getType(), this.configKey), -1);
+        if (tickRate > -1 && gameTime < this.lastRun + tickRate) {
+            return false;
+        }
+        // Nabulus end
         if (this.trigger(level, entity, gameTime)) {
             this.status = Behavior.Status.RUNNING;
+            this.lastRun = gameTime; // Nabulus - behavior tick rate
             return true;
         } else {
             return false;
diff --git a/net/minecraft/world/entity/ai/behavior/declarative/BehaviorBuilder.java b/net/minecraft/world/entity/ai/behavior/declarative/BehaviorBuilder.java
index 07e3d0070c015cbfb83a961eb70fa5a0b6ed1110..240b642169c0c6a1992cb769341eb59df81694ca 100644
--- a/net/minecraft/world/entity/ai/behavior/declarative/BehaviorBuilder.java
+++ b/net/minecraft/world/entity/ai/behavior/declarative/BehaviorBuilder.java
@@ -37,7 +37,15 @@ public class BehaviorBuilder<E extends LivingEntity, M> implements App<BehaviorB
         Function<BehaviorBuilder.Instance<E>, ? extends App<BehaviorBuilder.Mu<E>, Trigger<E>>> initializer
     ) {
         final BehaviorBuilder.TriggerWithResult<E, Trigger<E>> triggerWithResult = get((App<BehaviorBuilder.Mu<E>, Trigger<E>>)initializer.apply(instance()));
-        return new OneShot<E>() {
+        // Nabulus start - configurable tick rates
+        java.lang.StackTraceElement[] stackTraceElements = Thread.currentThread().getStackTrace();
+        String callingClass = null;
+        if (stackTraceElements.length > 2) {
+            callingClass = stackTraceElements[2].getClassName();
+        }
+        final String configKey = callingClass == null ? "unknown" : net.minecraft.world.entity.ai.behavior.Behavior.getConfigKey(callingClass);
+        // Nabulus end
+        return new OneShot<E>(configKey) { // Nabulus
             @Override
             public boolean trigger(ServerLevel level, E entity, long gameTime) {
                 Trigger<E> trigger = triggerWithResult.tryTrigger(level, entity, gameTime);
