From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Mon, 11 Mar 2024 18:40:31 +0930
Subject: [PATCH] Allow players to hide wither boss bar


diff --git a/net/minecraft/world/entity/boss/wither/WitherBoss.java b/net/minecraft/world/entity/boss/wither/WitherBoss.java
index afe43600c4976e01e61d716034a2823d50fb55cb..2553126e05d08eca41ee145e7a08f7e5456af7fb 100644
--- a/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -77,6 +77,21 @@ public class WitherBoss extends Monster implements RangedAttackMob {
     private static final TargetingConditions.Selector LIVING_ENTITY_SELECTOR = (entity, level) -> !entity.getType().is(EntityTypeTags.WITHER_FRIENDS)
         && entity.attackable();
     private static final TargetingConditions TARGETING_CONDITIONS = TargetingConditions.forCombat().range(20.0).selector(LIVING_ENTITY_SELECTOR);
+    // Nabulus start
+    public boolean hideBossBar = false;
+    private void updateBossBar() {
+        Component customName = getCustomName();
+        if (customName != null) {
+            String collapsedName = customName.tryCollapseToString();
+            hideBossBar = collapsedName != null && collapsedName.equalsIgnoreCase("nobossbar");
+        } else {
+            hideBossBar = false;
+        }
+        if (bossEvent != null && hideBossBar) {
+            bossEvent.removeAllPlayers();
+        }
+    }
+    // Nabulus end
 
     public WitherBoss(EntityType<? extends WitherBoss> entityType, Level level) {
         super(entityType, level);
@@ -131,6 +146,7 @@ public class WitherBoss extends Monster implements RangedAttackMob {
     @Override
     public void setCustomName(@Nullable Component name) {
         super.setCustomName(name);
+        updateBossBar(); // Nabulus
         this.bossEvent.setName(this.getDisplayName());
     }
 
@@ -401,7 +417,7 @@ public class WitherBoss extends Monster implements RangedAttackMob {
     @Override
     public void startSeenByPlayer(ServerPlayer player) {
         super.startSeenByPlayer(player);
-        this.bossEvent.addPlayer(player);
+        if (!hideBossBar) this.bossEvent.addPlayer(player); // Nabulus
     }
 
     @Override
