From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Mon, 11 Mar 2024 18:40:31 +0930
Subject: [PATCH] Allow players to hide wither boss bar


diff --git a/net/minecraft/world/entity/boss/wither/WitherBoss.java b/net/minecraft/world/entity/boss/wither/WitherBoss.java
index 419c29c003b3e690a89a14c0a9b3eb2ac3500175..4f93d1983618239db4fb02eb95955d9d5d42b63d 100644
--- a/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -82,6 +82,21 @@ public class WitherBoss extends Monster implements RangedAttackMob {
     private static final TargetingConditions.Selector LIVING_ENTITY_SELECTOR = (target, level) -> !target.is(EntityTypeTags.WITHER_FRIENDS)
         && target.attackable();
     private static final TargetingConditions TARGETING_CONDITIONS = TargetingConditions.forCombat().range(20.0).selector(LIVING_ENTITY_SELECTOR);
+    // Nabulus start
+    public boolean hideBossBar = false;
+    private void updateBossBar() {
+        Component customName = getCustomName();
+        if (customName != null) {
+            String collapsedName = customName.tryCollapseToString();
+            hideBossBar = collapsedName != null && collapsedName.equalsIgnoreCase("nobossbar");
+        } else {
+            hideBossBar = false;
+        }
+        if (bossEvent != null && hideBossBar) {
+            bossEvent.removeAllPlayers();
+        }
+    }
+    // Nabulus end
 
     public WitherBoss(final EntityType<? extends WitherBoss> type, final Level level) {
         super(type, level);
@@ -136,6 +151,7 @@ public class WitherBoss extends Monster implements RangedAttackMob {
     @Override
     public void setCustomName(final @Nullable Component name) {
         super.setCustomName(name);
+        updateBossBar(); // Nabulus
         this.bossEvent.setName(this.getDisplayName());
     }
 
@@ -403,7 +419,7 @@ public class WitherBoss extends Monster implements RangedAttackMob {
     @Override
     public void startSeenByPlayer(final ServerPlayer player) {
         super.startSeenByPlayer(player);
-        this.bossEvent.addPlayer(player);
+        if (!hideBossBar) this.bossEvent.addPlayer(player); // Nabulus
     }
 
     @Override
