From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Mon, 11 Mar 2024 18:40:31 +0930
Subject: [PATCH] Allow players to hide wither boss bar


diff --git a/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java b/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java
index 82921c56c49edb0ca07425da563aa4876d4e6fb1..83b2788fe8a5d12a4c73056110e3a922253c79d5 100644
--- a/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -93,6 +93,21 @@ public class WitherBoss extends Monster implements PowerableMob, RangedAttackMob
 
     public void setCanTravelThroughPortals(boolean canPortal) { this.canPortal = canPortal; }
     // Paper end
+    // Nabulus start
+    public boolean hideBossBar = false;
+    private void updateBossBar() {
+        Component customName = getCustomName();
+        if (customName != null) {
+            String collapsedName = customName.tryCollapseToString();
+            hideBossBar = collapsedName != null && collapsedName.equalsIgnoreCase("nobossbar");
+        } else {
+            hideBossBar = false;
+        }
+        if (bossEvent != null && hideBossBar) {
+            bossEvent.removeAllPlayers();
+        }
+    }
+    // Nabulus end
 
     public WitherBoss(EntityType<? extends WitherBoss> type, Level world) {
         super(type, world);
@@ -151,6 +166,7 @@ public class WitherBoss extends Monster implements PowerableMob, RangedAttackMob
     @Override
     public void setCustomName(@Nullable Component name) {
         super.setCustomName(name);
+        updateBossBar(); // Nabulus
         this.bossEvent.setName(this.getDisplayName());
     }
 
@@ -422,7 +438,7 @@ public class WitherBoss extends Monster implements PowerableMob, RangedAttackMob
     @Override
     public void startSeenByPlayer(ServerPlayer player) {
         super.startSeenByPlayer(player);
-        this.bossEvent.addPlayer(player);
+        if (!hideBossBar) this.bossEvent.addPlayer(player); // Nabulus
     }
 
     @Override
