From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sun, 27 Feb 2022 21:27:01 +0930
Subject: [PATCH] Separate creative levels


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 58591bf2f63b9c5e97d9ce4188dff3366968a178..86887f5be0281facd85c48fab9873cac20aec216 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -656,6 +656,7 @@ public class ServerPlayer extends Player {
             this.setPosRaw(position.x(), position.y(), position.z()); // Paper - don't register to chunks yet
         }
         this.gameMode.setLevel((ServerLevel) world);
+        this.chooseInventory(); // Nabulus
     }
     // CraftBukkit end
 
@@ -1251,6 +1252,7 @@ public class ServerPlayer extends Player {
 
                 // CraftBukkit end
                 this.setServerLevel(worldserver);
+                this.chooseInventory(); // Nabulus
                 this.connection.teleport(exit); // CraftBukkit - use internal teleport without event
                 this.connection.resetPosition();
                 worldserver.addDuringPortalTeleport(this);
@@ -1911,10 +1913,20 @@ public class ServerPlayer extends Player {
     }
 
     public void restoreFrom(ServerPlayer oldPlayer, boolean alive) {
+        // Nabulus start
+        this.restoreFrom(oldPlayer, alive, false);
+    }
+    public void restoreFrom(ServerPlayer oldPlayer, boolean alive, boolean toOrFromCreativeLevel) {
+        // Nabulus end
         this.wardenSpawnTracker = oldPlayer.wardenSpawnTracker;
         this.chatSession = oldPlayer.chatSession;
         this.gameMode.setGameModeForPlayer(oldPlayer.gameMode.getGameModeForPlayer(), oldPlayer.gameMode.getPreviousGameModeForPlayer());
         this.onUpdateAbilities();
+        // Nabulus start
+        this.normalInventory.replaceWith(oldPlayer.normalInventory);
+        this.creativeLevelInventory.replaceWith(oldPlayer.creativeLevelInventory);
+        this.usingNormalInventory = oldPlayer.usingNormalInventory;
+        // Nabulus end
         if (alive) {
             this.getInventory().replaceWith(oldPlayer.getInventory());
             this.setHealth(oldPlayer.getHealth());
@@ -1942,6 +1954,7 @@ public class ServerPlayer extends Player {
         this.seenCredits = oldPlayer.seenCredits;
         this.enteredNetherPosition = oldPlayer.enteredNetherPosition;
         this.chunkTrackingView = oldPlayer.chunkTrackingView;
+        if (toOrFromCreativeLevel) return; // Nabulus
         this.setShoulderEntityLeft(oldPlayer.getShoulderEntityLeft());
         this.setShoulderEntityRight(oldPlayer.getShoulderEntityRight());
         this.setLastDeathLocation(oldPlayer.getLastDeathLocation());
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 91feb12732564c90656da487664dbc12e55397fc..82fd3a5d4a93880f9895eb0f3b4fdb33df8fdcf9 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -843,8 +843,15 @@ public abstract class PlayerList {
         entityplayer.wonGame = false;
         // CraftBukkit end
 
+        // Nabulus start
+        boolean toOrFromCreativeLevel = ((net.minecraft.world.level.storage.PrimaryLevelData) worldserver.levelData).creativeLevel || ((net.minecraft.world.level.storage.PrimaryLevelData) entityplayer.level().levelData).creativeLevel;
+        if (toOrFromCreativeLevel) {
+            entityplayer.releaseLeftShoulderEntity();
+            entityplayer.releaseRightShoulderEntity();
+        }
+        // Nabulus end
         entityplayer1.connection = entityplayer.connection;
-        entityplayer1.restoreFrom(entityplayer, flag);
+        entityplayer1.restoreFrom(entityplayer, flag, toOrFromCreativeLevel); // Nabulus
         entityplayer1.setId(entityplayer.getId());
         entityplayer1.setMainArm(entityplayer.getMainArm());
         Iterator iterator = entityplayer.getTags().iterator();
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 0f2af6968f6bb47f5ef7e3f63cbfdc2f6df53999..2b474bf3d2c5afd9e1ef6f5922f35b04fe711b39 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3644,6 +3644,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
         }
         // Paper end - Fix item duplication and teleport issues
         if (this.level() instanceof ServerLevel && !this.isRemoved()) {
+            // Nabulus start
+            if (!this.level.equals(worldserver) && ((net.minecraft.world.level.storage.PrimaryLevelData) this.level().levelData).creativeLevel) {
+                return null;
+            }
+            // Nabulus end
             this.level().getProfiler().push("changeDimension");
             // CraftBukkit start
             // this.unRide();
@@ -3738,6 +3743,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
         if (destination == null) {
             return null;
         }
+        // Nabulus start
+        if (((net.minecraft.world.level.storage.PrimaryLevelData) this.level().levelData).creativeLevel) {
+            return null;
+        }
+        // Nabulus end
         boolean flag = this.level().getTypeKey() == LevelStem.END && destination.getTypeKey() == LevelStem.OVERWORLD; // fromEndToOverworld
         boolean flag1 = destination.getTypeKey() == LevelStem.END; // targetIsEnd
         // CraftBukkit end
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 5ca1f834f311a87323ced2578535e66efa14e47f..f7b88c3215635e60d1448bb6cf57410148bd088e 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -187,6 +187,40 @@ public abstract class Player extends LivingEntity {
     public boolean fauxSleeping;
     public int oldLevel = -1;
 
+    // Nabulus start
+    protected final Inventory normalInventory = new Inventory(this);
+    protected final Inventory creativeLevelInventory = new Inventory(this);
+    private boolean inventoriesLoaded = false;
+    public boolean usingNormalInventory = true;
+
+    public void chooseInventory() {
+        if (!inventoriesLoaded) {
+            return;
+        }
+        if (((net.minecraft.world.level.storage.PrimaryLevelData)this.level().levelData).creativeLevel) {
+            if (usingNormalInventory) {
+                this.normalInventory.replaceWith(this.inventory);
+                this.inventory.replaceWith(this.creativeLevelInventory);
+                usingNormalInventory = false;
+            }
+        } else {
+            if (!usingNormalInventory) {
+                this.creativeLevelInventory.replaceWith(this.inventory);
+                this.inventory.replaceWith(this.normalInventory);
+                usingNormalInventory = true;
+            }
+        }
+    }
+
+    public Inventory getNormalInventory() {
+        return usingNormalInventory ? this.inventory : this.normalInventory;
+    }
+
+    public Inventory getCreativeLevelInventory() {
+        return usingNormalInventory ? this.creativeLevelInventory : this.inventory;
+    }
+    // Nabulus end
+
     @Override
     public CraftHumanEntity getBukkitEntity() {
         return (CraftHumanEntity) super.getBukkitEntity();
@@ -200,6 +234,7 @@ public abstract class Player extends LivingEntity {
         this.lastDeathLocation = Optional.empty();
         this.setUUID(gameProfile.getId());
         this.gameProfile = gameProfile;
+        this.chooseInventory(); // Nabulus
         this.inventoryMenu = new InventoryMenu(this.inventory, !world.isClientSide, this);
         this.containerMenu = this.inventoryMenu;
         this.moveTo((double) pos.getX() + 0.5D, (double) (pos.getY() + 1), (double) pos.getZ() + 0.5D, yaw, 0.0F);
@@ -862,6 +897,15 @@ public abstract class Player extends LivingEntity {
 
         this.inventory.load(nbttaglist);
         this.inventory.selected = nbt.getInt("SelectedItemSlot");
+        // Nabulus start
+        this.normalInventory.load(nbttaglist);
+        this.normalInventory.selected = nbt.getInt("SelectedItemSlot");
+        ListTag creativeLevelInventoryTagList = nbt.getList("Nabulus.CreativeLevelInventory", 10);
+        this.creativeLevelInventory.load(creativeLevelInventoryTagList);
+        this.creativeLevelInventory.selected = nbt.getInt("Nabulus.CreativeLevelSelectedItemSlot");
+        this.inventoriesLoaded = true;
+        this.usingNormalInventory = true;
+        // Nabulus end
         this.sleepCounter = nbt.getShort("SleepTimer");
         this.experienceProgress = nbt.getFloat("XpP");
         this.experienceLevel = nbt.getInt("XpLevel");
@@ -901,8 +945,12 @@ public abstract class Player extends LivingEntity {
     public void addAdditionalSaveData(CompoundTag nbt) {
         super.addAdditionalSaveData(nbt);
         NbtUtils.addCurrentDataVersion(nbt);
-        nbt.put("Inventory", this.inventory.save(new ListTag()));
-        nbt.putInt("SelectedItemSlot", this.inventory.selected);
+        nbt.put("Inventory", this.getNormalInventory().save(new ListTag())); // Nabulus
+        nbt.putInt("SelectedItemSlot", this.getNormalInventory().selected); // Nabulus
+        // Nabulus start
+        nbt.put("Nabulus.CreativeLevelInventory", this.getCreativeLevelInventory().save(new ListTag()));
+        nbt.putInt("Nabulus.CreativeLevelSelectedItemSlot", this.getCreativeLevelInventory().selected);
+        // Nabulus end
         nbt.putShort("SleepTimer", (short) this.sleepCounter);
         nbt.putFloat("XpP", this.experienceProgress);
         nbt.putInt("XpLevel", this.experienceLevel);
diff --git a/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java b/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java
index 399da9d43aefbb95897df4697860d5bce5317152..58370c6c9988a404d2cdb5951bdae3e16f582085 100644
--- a/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -174,6 +174,7 @@ public class LevelStorageSource {
         Lifecycle lifecycle = worlddimensions_b.lifecycle().add(registryManager.allRegistriesLifecycle());
         PrimaryLevelData worlddataserver = PrimaryLevelData.parse(dynamic1, worldsettings, worlddimensions_b.specialWorldProperty(), generatorsettings.options(), lifecycle);
         worlddataserver.pdc = ((Dynamic<Tag>) dynamic1).getElement("BukkitValues", null); // CraftBukkit - Add PDC to world
+        worlddataserver.creativeLevel = ((Dynamic<Tag>) dynamic1).get("Nabulus.CreativeLevel").asBoolean(false); // Nabulus
 
         return new LevelDataAndDimensions(worlddataserver, worlddimensions_b);
     }
diff --git a/src/main/java/net/minecraft/world/level/storage/PrimaryLevelData.java b/src/main/java/net/minecraft/world/level/storage/PrimaryLevelData.java
index 59ba982dc96ce47e47399514e8f74d2b972dbe1e..b79d49b959bb4e90e7773fe9018f36f9c01ff2df 100644
--- a/src/main/java/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/src/main/java/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -96,6 +96,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
     public Registry<LevelStem> customDimensions;
     private ServerLevel world;
     protected Tag pdc;
+    public boolean creativeLevel;
 
     public void setWorld(ServerLevel world) {
         if (this.world != null) {
@@ -261,6 +262,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
 
         levelNbt.putString("Bukkit.Version", Bukkit.getName() + "/" + Bukkit.getVersion() + "/" + Bukkit.getBukkitVersion()); // CraftBukkit
         this.world.getWorld().storeBukkitValues(levelNbt); // CraftBukkit - add pdc
+        levelNbt.putBoolean("Nabulus.CreativeLevel", this.creativeLevel); // Nabulus
     }
 
     private static ListTag stringCollectionToTag(Set<String> strings) {
