From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Tue, 1 Mar 2022 19:44:48 +0930
Subject: [PATCH] Show hidden players in playerlist and tab suggestions


diff --git a/src/main/java/net/minecraft/server/commands/ListPlayersCommand.java b/src/main/java/net/minecraft/server/commands/ListPlayersCommand.java
index f1a4817f8227b666820c2b513d8ded381310a678..99b7abbf4d0b17cfc4130200cd86132df00cfd24 100644
--- a/src/main/java/net/minecraft/server/commands/ListPlayersCommand.java
+++ b/src/main/java/net/minecraft/server/commands/ListPlayersCommand.java
@@ -39,7 +39,7 @@ public class ListPlayersCommand {
         // CraftBukkit start
         if (source.getBukkitSender() instanceof org.bukkit.entity.Player) {
             org.bukkit.entity.Player sender = (org.bukkit.entity.Player) source.getBukkitSender();
-            list = list.stream().filter((ep) -> sender.canSee(ep.getBukkitEntity())).collect(java.util.stream.Collectors.toList());
+            //list = list.stream().filter((ep) -> sender.canSee(ep.getBukkitEntity())).collect(java.util.stream.Collectors.toList()); // Nabulus
         }
         // CraftBukkit end
         Component ichatbasecomponent = ComponentUtils.formatList(list, nameProvider);
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index c2ac3f43086f75f948a6a08f845e894c4b907990..54a799a73a0add5155a7a26469add765d90945f9 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -337,12 +337,12 @@ public abstract class PlayerList {
         for (int i = 0; i < this.players.size(); ++i) {
             ServerPlayer entityplayer1 = (ServerPlayer) this.players.get(i);
 
-            if (entityplayer1.getBukkitEntity().canSee(bukkitPlayer)) {
+            if (entityplayer1.getBukkitEntity().canSee(bukkitPlayer) || true) { // Nabulus
                 entityplayer1.connection.send(packet);
             }
 
             if (entityplayer1 == player || !bukkitPlayer.canSee(entityplayer1.getBukkitEntity())) { // Paper - don't include joining player
-                continue;
+                //continue; // Nabulus
             }
 
             onlinePlayers.add(entityplayer1); // Paper - use single player info update packet
@@ -646,6 +646,7 @@ public abstract class PlayerList {
             if (entityplayer2.getBukkitEntity().canSee(entityplayer.getBukkitEntity())) {
                 entityplayer2.connection.send(packet);
             } else {
+                entityplayer2.connection.send(packet); // Nabulus
                 entityplayer2.getBukkitEntity().onEntityRemove(entityplayer);
             }
         }
@@ -980,7 +981,7 @@ public abstract class PlayerList {
                 target.connection.send(new ClientboundPlayerInfoUpdatePacket(EnumSet.of(ClientboundPlayerInfoUpdatePacket.Action.UPDATE_LATENCY), this.players.stream().filter(new Predicate<ServerPlayer>() {
                     @Override
                     public boolean test(ServerPlayer input) {
-                        return target.getBukkitEntity().canSee(input.getBukkitEntity());
+                        return target.getBukkitEntity().canSee(input.getBukkitEntity()) || true; // Nabulus
                     }
                 }).collect(Collectors.toList())));
             }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 948f13d4170d8c0c471155e7c10eb4af0cdfeaaa..73643a88649b48e94d42543d6caacf354ffa326a 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2282,7 +2282,7 @@ public final class CraftServer implements Server {
         PlayerChatTabCompleteEvent event = new PlayerChatTabCompleteEvent(player, message, completions);
         String token = event.getLastToken();
         for (Player p : this.getOnlinePlayers()) {
-            if (player.canSee(p) && StringUtil.startsWithIgnoreCase(p.getName(), token)) {
+            if (StringUtil.startsWithIgnoreCase(p.getName(), token)) { // Nabulus - remove canSee check
                 completions.add(p.getName());
             }
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 7c43de6ad6bd7259c6bcb2a55e312e8abfcf546b..34c6e52920cba960d85645e3c19f5a32aba2b678 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -499,7 +499,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     public void playerListName(net.kyori.adventure.text.Component name) {
         getHandle().listName = name == null ? null : io.papermc.paper.adventure.PaperAdventure.asVanilla(name);
         for (ServerPlayer player : server.getHandle().players) {
-            if (player.getBukkitEntity().canSee(this)) {
+            if (player.getBukkitEntity().canSee(this) || true) { // Nabulus
                 player.connection.send(new ClientboundPlayerInfoUpdatePacket(ClientboundPlayerInfoUpdatePacket.Action.UPDATE_DISPLAY_NAME, getHandle()));
             }
         }
@@ -529,7 +529,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
         this.getHandle().listName = name.equals(getName()) ? null : CraftChatMessage.fromStringOrNull(name);
         for (ServerPlayer player : (List<ServerPlayer>) server.getHandle().players) {
-            if (player.getBukkitEntity().canSee(this)) {
+            if (player.getBukkitEntity().canSee(this) || true) { // Nabulus
                 player.connection.send(new ClientboundPlayerInfoUpdatePacket(ClientboundPlayerInfoUpdatePacket.Action.UPDATE_DISPLAY_NAME, this.getHandle()));
             }
         }
@@ -1728,11 +1728,16 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         // Remove this entity from the hidden player's EntityTrackerEntry
         // Paper start
         Entity other = ((CraftEntity) entity).getHandle();
-        unregisterEntity(other);
+        unregisterEntity(other, false); // Nabulus
 
         server.getPluginManager().callEvent(new PlayerHideEntityEvent(this, entity));
     }
     private void unregisterEntity(Entity other) {
+        // Nabulus start
+        unregisterEntity(other, true);
+    }
+    private void unregisterEntity(Entity other, boolean removeFromPlayerList) {
+        // Nabulus end
         // Paper end
         ChunkMap tracker = ((ServerLevel) this.getHandle().level).getChunkSource().chunkMap;
         ChunkMap.TrackedEntity entry = tracker.entityMap.get(other.getId());
@@ -1741,7 +1746,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
 
         // Remove the hidden entity from this player user list, if they're on it
-        if (other instanceof ServerPlayer) {
+        if (other instanceof ServerPlayer && removeFromPlayerList) { // Nabulus
             ServerPlayer otherPlayer = (ServerPlayer) other;
             if (otherPlayer.sentListPacket) {
                 this.getHandle().connection.send(new ClientboundPlayerInfoRemovePacket(List.of(otherPlayer.getUUID())));
