From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sat, 6 Aug 2022 13:21:50 +0930
Subject: [PATCH] Use system messages for player chat until Velocity support
 added


diff --git a/src/main/java/net/minecraft/network/chat/OutgoingChatMessage.java b/src/main/java/net/minecraft/network/chat/OutgoingChatMessage.java
index 74cf1c043beef03cfd5adf481414a5ee78bef2a6..5e526c9d6f2233e65f0586d5ad933d5dc4406408 100644
--- a/src/main/java/net/minecraft/network/chat/OutgoingChatMessage.java
+++ b/src/main/java/net/minecraft/network/chat/OutgoingChatMessage.java
@@ -45,7 +45,7 @@ public interface OutgoingChatMessage {
             PlayerChatMessage playerChatMessage = this.message.filter(filterMaskEnabled);
             playerChatMessage = unsigned != null ? playerChatMessage.withUnsignedContent(unsigned) : playerChatMessage; // Paper
             if (!playerChatMessage.isFullyFiltered()) {
-                sender.connection.sendPlayerChatMessage(playerChatMessage, params);
+                sender.connection.sendDisguisedChatMessage(playerChatMessage.decoratedContent(), params); // Nabulus
             }
 
         }
