From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sun, 12 Jun 2022 10:24:09 +0930
Subject: [PATCH] Decrease AcquirePoi update rate


diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/AcquirePoi.java b/src/main/java/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
index 43243537b765a2d270be6de3f053fea77ff67d18..c2bafd7ce9ff99bb42d1603aae9383405d84e97f 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
@@ -71,7 +71,7 @@ public class AcquirePoi extends Behavior<PathfinderMob> {
 
     @Override
     protected void start(ServerLevel world, PathfinderMob entity, long time) {
-        this.nextScheduledStart = time + 20L + (long)world.getRandom().nextInt(20);
+        this.nextScheduledStart = time + 100L + (long)world.getRandom().nextInt(20);
         PoiManager poiManager = world.getPoiManager();
         this.batchCache.long2ObjectEntrySet().removeIf((entry) -> {
             return !entry.getValue().isStillValid(time);
