From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Fri, 10 Feb 2023 19:54:12 +0930
Subject: [PATCH] Add silent chest flag to players


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index f7b88c3215635e60d1448bb6cf57410148bd088e..df163093bdc3c7da9cd8ba2159b06b051bc99302 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -188,6 +188,7 @@ public abstract class Player extends LivingEntity {
     public int oldLevel = -1;
 
     // Nabulus start
+    public boolean silentChests = false;
     protected final Inventory normalInventory = new Inventory(this);
     protected final Inventory creativeLevelInventory = new Inventory(this);
     private boolean inventoriesLoaded = false;
diff --git a/src/main/java/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java b/src/main/java/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
index 7eccacb978f047a4774e58b11bc3f9ab3959b049..debc7165ea0af4ef4cb983e07c0a9f20c83696f8 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
@@ -40,6 +40,7 @@ public abstract class ContainerOpenersCounter {
     protected abstract boolean isOwnContainer(Player player);
 
     public void incrementOpeners(Player player, Level world, BlockPos pos, BlockState state) {
+        if (player.silentChests) return; // Nabulus
         int oldPower = Math.max(0, Math.min(15, this.openCount)); // CraftBukkit - Get power before new viewer is added
         int i = this.openCount++;
 
@@ -63,6 +64,7 @@ public abstract class ContainerOpenersCounter {
     }
 
     public void decrementOpeners(Player player, Level world, BlockPos pos, BlockState state) {
+        if (player.silentChests) return; // Nabulus
         int oldPower = Math.max(0, Math.min(15, this.openCount)); // CraftBukkit - Get power before new viewer is added
         if (this.openCount == 0) return; // Paper - Prevent ContainerOpenersCounter openCount from going negative
         int i = this.openCount--;
@@ -92,7 +94,7 @@ public abstract class ContainerOpenersCounter {
         float f = 5.0F;
         AABB axisalignedbb = new AABB((double) ((float) i - 5.0F), (double) ((float) j - 5.0F), (double) ((float) k - 5.0F), (double) ((float) (i + 1) + 5.0F), (double) ((float) (j + 1) + 5.0F), (double) ((float) (k + 1) + 5.0F));
 
-        return world.getEntities(EntityTypeTest.forClass(Player.class), axisalignedbb, this::isOwnContainer).size();
+        return world.getEntities(EntityTypeTest.forClass(Player.class), axisalignedbb, player -> this.isOwnContainer(player) && !player.silentChests).size(); // Nabulus - add silent chest check
     }
 
     public void recheckOpeners(Level world, BlockPos pos, BlockState state) {
diff --git a/src/main/java/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
index 1fa22445a4ecc8c08dbcf0cc6bd39dc5003604c4..d31fb10742f6570c248e80467495f796f8aefa5c 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
@@ -201,7 +201,7 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
 
     @Override
     public void startOpen(Player player) {
-        if (!this.remove && !player.isSpectator()) {
+        if (!this.remove && !player.isSpectator() && !player.silentChests) { // Nabulus
             if (this.openCount < 0) {
                 this.openCount = 0;
             }
@@ -219,7 +219,7 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
 
     @Override
     public void stopOpen(Player player) {
-        if (!this.remove && !player.isSpectator()) {
+        if (!this.remove && !player.isSpectator() && !player.silentChests) { // Nabulus
             --this.openCount;
             if (this.opened) return; // CraftBukkit - only animate if the ShulkerBox hasn't been forced open already by an API call.
             this.level.blockEvent(this.worldPosition, this.getBlockState().getBlock(), 1, this.openCount);
