From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sun, 12 Feb 2023 10:10:22 +0930
Subject: [PATCH] Fix pardoning uncached game profiles


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java b/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java
index 535a571d140dd5932dbd503c540f988c421f0627..6b4cddda38a0828b2677bb5afe37b6c8ae0dc8b5 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java
@@ -92,6 +92,19 @@ public class CraftProfileBanList implements org.bukkit.BanList {
         Preconditions.checkArgument(target != null, "Target cannot be null");
 
         GameProfile profile = this.getProfile(target);
+        // Nabulus start
+        if (profile == null) {
+            java.util.List<GameProfile> toPardon = new java.util.ArrayList<>();
+            for (UserBanListEntry banListEntry : list.getEntries()) {
+                if (banListEntry == null || banListEntry.getUser() == null || banListEntry.getUser().getId() == null) continue;
+                if (banListEntry.getUser().getId().toString().equalsIgnoreCase(target)) {
+                    toPardon.add(banListEntry.getUser());
+                }
+            }
+            toPardon.forEach(this.list::remove);
+            return;
+        }
+        // Nabulus end
         this.list.remove(profile);
     }
 
