From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sat, 16 Dec 2023 15:18:46 +0930
Subject: [PATCH] Don't show smoke for silent fireworks


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
index 7005b3a26e9b5d79064981a4a41bee21b65a9fc3..21fdd9e3db1ed71414bd30e54b75b41bafe51786 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -170,7 +170,7 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
         }
 
         ++this.life;
-        if (this.level().isClientSide && this.life % 2 < 2) {
+        if (this.level().isClientSide && this.life % 2 < 2 && !this.isSilent()) { // Nabulus - don't show firework smoke for silent fireworks
             this.level().addParticle(ParticleTypes.FIREWORK, this.getX(), this.getY(), this.getZ(), this.random.nextGaussian() * 0.05D, -this.getDeltaMovement().y * 0.5D, this.random.nextGaussian() * 0.05D);
         }
 
